### ISSUE TITLE
Defining and retrieving additional keywords

### ISSUE THREAD

**Issue by aeneasr** [xeipuuv/gojsonschema#244]
I would like to add additional supported keywords to my JSON Schema, for example:

```json
{
"type": "string",
"foo": "bar"
}
```

Now, I want to be able to check if a field has set keyword "foo" and also get its value (similar to 
`IsFormat(input interface{}) bool`)

Since we do already have the ability to add custom `FormatCheckers`, I took it that it should/would be possible to add other keywords with custom logic too, for example:

```
type f struct {}

func (a f) Call(value interface{}, input interface{}) error {
}

gojsonschema.Keywords.Add("foo", f)
```

Is that something you'd accept in this repository? I'm happy to cook up a PR but would definitely need some guidance :)

**Comment by johandorland** [comment]
Yes and no. The current code is totally not designed to expose it's internals to the outside world. An interface like the one you suggested only allows for very basic functionality as it doesn't tie into the rest of the gojsonschema code (i.e. You can't have a keyword that at some point contains a JSON schema as you'd need access to the internal functions to parse it recursively, etc). 

The good news is that the upcoming draft-08 of JSON schema has support for vocabularies, which basically requires implementations  to provide APIs to users to define their own keywords. It does mean however `gojsonschema` will need a thorough internal rewrite because it has to expose a clean API that works with annotation collection.

Although it'd be possible to expose a simple API like you suggested it would be limited, ugly and I really don't want to support anything like that if there is also a proper solution.

Feel free to implement `$vocabulary` and annotation collection support, but I should warn you that it will be a lot of work and requires thorough knowledge of the JSON schema specification. I intend to do it at some point in the future myself, but for me this is only a hobby project at the moment.

**Comment by aeneasr** [comment]
Thank you for the prompt response! I'm definitely not capable of implementing draft-8, I took a quick glance at the proposal and it seems like there's a lot of new things going on there. I took another glance at the current source code and I think it would be fairly easy to add a hook with all of the context (basically within the recursion, before or after validation), that would allow to do what I want to do with fairly few code changes. I'm pretty sure that's not something you'd like to have in this library though. However, I do think that it could make sense to export some of the methods and structs so that one could theoretically just wrap those, or simply iterate over the JSONSchema him/herself.

While we do not have an Open Source Sponsorship program at [Ory](http://github.com/ory) yet, we will have one in the future. Supporting draft-8 with vocabulary support would be very useful for us so maybe there's some way (once the draft is finished) to sponsor the implementation.

On a side node, assuming I were to I was to add custom fields/keys to the JSON Schema definition such as:

```json
{
  "type": "string",
  "foo": "bar"
}
```

If I understood correctly, I could write a `$vocabulary` that extends the draft-7 semantics (wrt to type, enum, ...) and adds certain keys so that the above code would be valid, right?

```json
{
 "$vocabulary": {
         "https://json-schema.org/draft-08/vocabularies/core": true,
         "https://my-custom-schema/with/key/foo": true
 }
  "type": "string",
  "foo": "bar"
}
```

I just want to make sure that if we add those extra keys, we can make that draft-8 compatible without having to rewrite parts (or all of) the schema.

**Comment by johandorland** [comment]
Depending on how you'd do it is possible, but the internal methods were never designed for it.  I feel like adding an interface that exposes those methods is only going to create some sort of Frankenstein interface as later on it will have to coexist with the `$vocabulary` interface. 

Your interpretation of `$vocabulary` is correct, so you could safely fork this library, implement custom keywords (which would be even easier to implement than having to expose the internals first), and then later switch back after proper `$vocabulary` support is implemented here.

**Comment by aeneasr** [comment]
Ok perfect, I'll go along with that approach then for now! I agree with your sentiment wrt an interface and I also think that proper `$vocabulary` with a clean API makes much more sense than trying to do this in a hacky fashion right now.

I'll keep monitoring the draft-8 status and, if you're up for it, cycle back regarding sponsoring a release towards draft-8 once it moves out of draft status :)
